plugins/com.google.cloud.tools.eclipse.util/src/com/google/cloud/tools/eclipse/util/MavenUtils.java

```diff
@@ -63,12 +82,62 @@ public static boolean hasMavenNature(IProject project) {
     }
   }
 
-  public static Artifact resolveArtifact(IProgressMonitor monitor, String groupId,
-      String artifactId, String type, String version, String classifier,
-      List<ArtifactRepository> repositories) throws CoreException {
-    Artifact artifact = MavenPlugin.getMaven().resolve(groupId, artifactId, version, type,
-        classifier, repositories, monitor);
-    return artifact;
+  public static Artifact resolveArtifact(
+      String groupId,
+      String artifactId,
+      String type,
+      String version,
+      String classifier,
+      List<ArtifactRepository> repositories,
+      IProgressMonitor monitor)
+      throws CoreException {
+    return runWithRule(
+        monitor,
+        (context, system, progress) -> {
+          Artifact artifact =
+              MavenPlugin.getMaven()
+                  .resolve(groupId, artifactId, version, type, classifier, repositories, progress);
+          return artifact;
+        });
+  }
+
+  /**
+   * Perform some Maven-related action (that may throw an exception), ensuring that the given {@link
+   * ISchedulingRule scheduling rule} is held.
+   *
+   * @param ruleMonitor a progress monitor to be used when waiting to obtain the rule
+   * @param rule the rule that must be held
+   */
+  public static <T, E extends Throwable> T runWithRule(
+      IProgressMonitor monitor, ExceptionalCallableWithProgress<T> supplier) throws CoreException {
+    SubMonitor progress = SubMonitor.convert(monitor, 10);
+    ISchedulingRule rule = mavenResolvingRule();
+    boolean acquireRule = Job.getJobManager().currentRule() == null;
+    if (acquireRule) {
+      Job.getJobManager().beginRule(rule, progress.split(2));
+    }
+    Verify.verify(
```

Comment:
Let's push this into `try`, so that we `endRule()` when `verify()` fails for any reason.